### PR TITLE
Support for Cucumber event listeners for TypeScript

### PR DESCRIPTION
**Pull request by jan-molak**
This PR adds types for Cucumber event listeners and is based on the [Serenity/JS implementation](https://github.com/jan-molak/serenity-js/blob/3e733369f4fd456239c4d97ef20ae4447e407799/src/serenity-cucumber/cucumber.d.ts).

---

Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/cucumber/cucumber-js/pull/130
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: cucumber/cucumber-tests.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
-/// <reference types="assert" />
 import cucumber = require("cucumber");
 
+var assert = {deepEqual: function (a: Object, b: Object) {
```
Comment (andy-ms) [inline cucumber/cucumber-tests.ts]: Why this change? The tests won't actually run, just be compiled. So why not use the already existing definition?

File: cucumber/index.d.ts [hunk 2]
```diff
@@ -70,5 +70,139 @@ declare namespace cucumber {
...
+	export function Listener(): EventListener;
+
+	export namespace events {
+
+		interface Event {
+			getName(): string;
+			getPayloadItem(name: string): EventPayload;
+		}
+
+		interface EventPayload {
```
Comment (andy-ms) [inline cucumber/index.d.ts]: An empty interface is equivalent to `{}`.

File: cucumber/index.d.ts [hunk 3]
```diff
@@ -1,7 +1,6 @@
 // Type definitions for cucumber-js
 // Project: https://github.com/cucumber/cucumber-js
 // Definitions by: Abraão Alves <https://github.com/abraaoalves>
-// Definitions by: Jan Molak <https://github.com/jan-molak>
```
Comment (andy-ms) [inline cucumber/index.d.ts]: You can add yourself after a comma: `// Definitions by: Abraão Alves <https://github.com/abraaoalves>, Jan Molak <https://github.com/jan-molak>`

File: cucumber/cucumber-tests.ts (hunk 1 above)
Comment (andy-ms) [inline cucumber/cucumber-tests.ts]: Fixed by #13602, please merge that into your branch.

File: cucumber/index.d.ts (hunk 2 above)
Comment (jan-molak) [inline cucumber/index.d.ts]: Well, yes and no :-) It seems like a bit more context is needed here.

To notify a `Listener` of different phases in the test scenario lifecycle, Cucumber emits events which contain a field of "type" `EventPayload`, which value you can retrieve via the `getEventPayload(name)` method on the `Event` object. Regretfully, those different types of payloads have nothing in common API-wise, only the fact that they're passed as part of the event. 

I used the marker interface to:
- satisfy the signature to `Event::getPayloadItem(name: string): EventPayload` 
- highlight that the method can return all those different kinds of events

An alternative approach would be to define the method signature as returning `any`, remove the marker interface and disassociate the payloads from each other since they don't have anything in common design-wise, but I didn't find it convincing. I could be missing other alternatives though so I'm open to suggestions!

The nice thing about having both the method signature and the payloads typed though is that you can define a listener like this:

```typescript

let listener = Object.assign(cucumber.Listener(), {
    handleBeforeScenarioEvent,
    // ...
});

function handleBeforeScenarioEvent (scenario: cucumber.events.ScenarioPayload, callback: () => void) {
    
    // stuff

    callback();
}
```

File: cucumber/index.d.ts (hunk 3 above)
Comment (jan-molak) [inline cucumber/index.d.ts]: Nah, that's cool. But it's good to know, thanks ;-)

File: cucumber/cucumber-tests.ts (hunk 1 above)
Comment (jan-molak) [inline cucumber/cucumber-tests.ts]: Awesome, give me a sec :-)

File: cucumber/index.d.ts (hunk 3 above)
Comment (jan-molak) [inline cucumber/index.d.ts]: Cool, thanks :-)

File: cucumber/index.d.ts (hunk 2 above)
Comment (andy-ms) [inline cucumber/index.d.ts]: Could you provide an example of using the `EventPayload` type?
To clarify, an empty interface really is equivalent to `{}` and doesn't "mark" anything. So `const foo: EventPayload = 0;` would compile without error.
If you need to specify that only the things you've designated here as `EventPayload`s qualify, you could use a union type: `type EventPayload = FeaturesPayload | FeatueresResultPayload | ... | StepResultPayload`.

Comment (jan-molak) [inline cucumber/index.d.ts]: Ah yes, that's a good point. A union type might actually work better here, thanks for your suggestion. I'll update the PR tonight.

**Comment by andy-ms** [comment]
@jan-molak Please fix merge conflicts.
